### DISCUSSION TITLE
Base architecture

### DISCUSSION THREAD

**Discussion by rockerbacon** [rockerbacon/ice-cream-machine-core#3] [Technical]
We need to define the base system architecture. The architecture should:
1. Provide an independent back-end
2. Be game agnostic
3. Interact with independently developed components in order to fulfil game-specific requirements

**Comment by nlaranjeira01** [comment]
1. We will separate the system in a client and server manner. The server will be the **core** of the system, responsible for the communication with the OS and the mod scripts. The client will be responsible for the **UI** in which the users will be able to interact with the core.
2. and 3. The system will be able to execute the necessary steps to install a mod based on mod scripts. There will be a standard library available to be used on the script components. This standard library will be responsible for operations that are OS-sensitive (e.g. filesystem operations).

**Reply by rockerbacon** [reply]
From what I understood, you're suggesting the following components:
- **UI**: user-facing interface for indirect interaction with the **Core**
- **Core**: main system functionality, requires an external **Module** for each game being managed
- **Module**: provides game-specific functionality, ensuring the **Core** remains game agnostic

**Reply by rockerbacon** [reply]
> There will be a standard library available to be used on the script components. This standard library will be responsible for operations that are OS-sensitive (e.g. filesystem operations)

This is something I don't fully agree with. I believe modules shouldn't directly interact with the external world (OS), in order to reduce module responsibilities. The less modules actively do, the easier module development is.
We may be able to restrict module responsibilities, so that they mostly serve as information authorities rather than active actors.

**Reply by nlaranjeira01** [reply]
Agreed.
Modules should more restricted and the external world (OS) interaction will be the core's responsibility.

**Comment by rockerbacon** [comment]
### Summary of the final decision
The architecture will contain 3 main components:
1. **UI**: the user-facing interface, through which users can access all available functionalities
2. **Core**: responsible for providing the necessary functionalities and interacting with the external world. This component works in a game agnostic manner, requiring additional software modules for each game
3. **Module**: provides game-specific information to the **Core**
